### PR TITLE
[feat] Enable TP and batching for PixtralVisionModel / Mistral3VLM

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Summary by CodeRabbit

* **New Features**
  * Improved support for batched and variable-sized image inputs in vision-language models, enabling more efficient processing.
  * Enhanced tensor parallelism support in vision transformer models, allowing for better scalability and performance.

* **Bug Fixes**
  * Corrected sequence length calculations in attention metadata for vision models to ensure accurate handling of batches.

* **Chores**
  * Added necessary dependencies to support image padding operations.
  * Introduced MPI-based parallel testing infrastructure for tensor parallelism validation in vision transformer models.

# [5382078][5389014][feat] Enable TP and batching for PixtralVisionModel / Mistral3VLM

## Description

This commit implements tensor parallelism and batching support for the `PixtralVisionModel` (used by `Mistral3VLM`).

For TP, the gating issue was that there was no TP supported implementation of the `torch.nn.Conv2D` layer. Since there is just a single `Conv2D` layer at the very beginning of the vision encoder, this PR proposes to run that layer on all ranks, and slice its outputs instead, while the remainder of the vision tower (vision attention layers, etc.) all function with TP as they usually would.

For batching, the padding from HF's original implementation (which unfortunately only supports numpy) is functionally replicated using `torchvision`.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_pixtral.py [hunk 1]
```diff
@@ -235,19 +236,18 @@ def forward(
     def load_weights(self, weights):
         modeling_utils._load_weights_impl(self, weights)
 
-    def _prepare_attn_metadata(self, batch_size: int, seq_len: int):
+    def _prepare_attn_metadata(self, batch_size: int, seq_lengths: List[int]):
         request_ids = list(range(1, batch_size + 1))
-        prompt_lens = [seq_len] * batch_size
         attn_metadata = self._metadata_cls(
-            seq_lens=torch.tensor([seq_len] * batch_size, dtype=torch.int),
+            seq_lens=torch.tensor(seq_lengths, dtype=torch.int),
```
[the remark this answers is not in the document]
Comment (symphonylyh) [reply tensorrt_llm/_torch/models/modeling_pixtral.py:242]: can you also do a favor to remove the `*batch_size` at this [line](https://github.com/NVIDIA/TensorRT-LLM/blob/afd00fe9ebb0392d66747ec94bca895db5ff1a46/tensorrt_llm/_torch/models/modeling_clip.py#L205)?

File: tensorrt_llm/_torch/models/modeling_pixtral.py [hunk 2]
```diff
@@ -109,7 +109,8 @@ def __init__(self, config: model_config_lib.ModelConfig[transformers.PixtralVisi
         self.layers = torch.nn.ModuleList()
         for i in range(config.pretrained_config.num_hidden_layers):
             self.layers.append(PixtralAttentionLayer(config=config, layer_idx=i))
-        self._head_dim = config.pretrained_config.head_dim
+        # @reviewer: Should we be checking if `tp_size` divides `head_dim` in the first place?
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_pixtral.py:112]: Because I'm an airhead and meant to shard the `num_heads`, not the `head_dim` 😅 Will invert the change and test it out.

**Review by amukkara — COMMENTED** [review]
@2ez4bz can you add a TP=2 test case?

**Review by amukkara — APPROVED** [review]

**Review by symphonylyh — APPROVED** [review]

**Review by yechank-nvidia — APPROVED** [review]
LGTM. One question below.
